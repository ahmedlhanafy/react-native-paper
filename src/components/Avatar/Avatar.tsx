// @component ./AvatarIcon.tsx
export { default as Icon, AvatarIconProps } from './AvatarIcon';

// @component ./AvatarImage.tsx
export { default as Image, AvatarImageProps } from './AvatarImage';

// @component ./AvatarText.tsx
export { default as Text, AvatarTextProps } from './AvatarText';
