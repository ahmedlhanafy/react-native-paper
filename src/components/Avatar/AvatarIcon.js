/* @flow */

import * as React from 'react';
import { View, StyleSheet } from 'react-native';
import color from 'color';
import Icon from '../Icon';
import { withTheme } from '../../core/theming';
import { white } from '../../styles/colors';
import type { Theme } from '../../types';
import type { IconSource } from './../Icon';

type Props = {
  /**
   * Icon to display for the `Avatar`.
   */
  icon: IconSource,
  /**
   * Size of the avatar.
   */
  size: number,
  /**
   * Custom color.
   */
  color?: string,
  style?: any,
  /**
   * @optional
   */
  theme: Theme,
};

/**
 * Avatars can be used to represent people in a graphical way.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/avatar-icon.png" />
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Avatar } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Avatar.Icon size={24} icon="folder" />
 * );
 * ```
 */
class Avatar extends React.Component<Props> {
  static displayName = 'Avatar.Icon';

  static defaultProps = {
    size: 64,
  };

  render() {
    const { icon, size, style, theme } = this.props;
    const { colors } = theme;

    const { backgroundColor = colors.primary } =
      StyleSheet.flatten(style) || {};
    const textColor =
      this.props.color || color(backgroundColor).light()
        ? white
        : 'rgba(0, 0, 0, .54)';

    return (
      <View
        style={[
          {
            width: size,
            height: size,
            borderRadius: size / 2,
            backgroundColor,
          },
          style,
        ]}
      >
        <View style={[styles.icon, { height: size }]}>
          <Icon source={icon} color={textColor} size={size * 0.75} />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  icon: {
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default withTheme(Avatar);
