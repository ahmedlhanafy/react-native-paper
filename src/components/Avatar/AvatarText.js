/* @flow */

import * as React from 'react';
import { View, StyleSheet } from 'react-native';
import color from 'color';
import Text from '../Typography/Text';
import { withTheme } from '../../core/theming';
import { white } from '../../styles/colors';
import type { Theme } from '../../types';

type Props = {
  /**
   * Text of the `Avatar`.
   */
  label?: string,
  /**
   * Size of the avatar.
   */
  size: number,
  /**
   * Custom color.
   */
  color?: string,
  /**
   * Custom background color.
   */
  backgroundColor?: string,
  style?: any,
  /**
   * @optional
   */
  theme: Theme,
};

/**
 * Avatars can be used to represent people in a graphical way.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/avatar-text.png" />
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Avatar } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Avatar.Text size={24} label="JD" />
 * );
 * ```
 */
class AvatarText extends React.Component<Props> {
  static displayName = 'Avatar.Text';

  static defaultProps = {
    label: '',
    size: 64,
  };

  render() {
    const { label, size, style, theme } = this.props;
    const { colors } = theme;

    const backgroundColor = this.props.backgroundColor || colors.primary;
    const textColor =
      this.props.color || color(backgroundColor).light()
        ? white
        : 'rgba(0, 0, 0, .54)';

    return (
      <View
        style={[
          {
            width: size,
            height: size,
            borderRadius: size / 2,
            backgroundColor,
          },
          style,
        ]}
      >
        <Text
          style={[
            styles.text,
            {
              color: textColor,
              fontSize: size * 0.5,
              height: size,
              lineHeight: size,
            },
          ]}
          numberOfLines={1}
        >
          {label}
        </Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  text: {
    textAlign: 'center',
  },
});

export default withTheme(AvatarText);
