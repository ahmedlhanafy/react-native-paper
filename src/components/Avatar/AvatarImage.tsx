import * as React from 'react';

import {
  Image,
  ImageSourcePropType,
  StyleProp,
  StyleSheet,
  View,
  ViewStyle,
} from 'react-native';

import AvatarText from './AvatarText';
import { Theme } from '../../types';
import { withTheme } from '../../core/theming';

const defaultSize = 64;

type Props = React.ComponentPropsWithRef<typeof View> & {
  /**
   * Image to display for the `Avatar`.
   */
  source: ImageSourcePropType;
  /**
   * Alternative text to avatar.
   */
  alt?: string;
  /**
   * Size of the avatar.
   */
  size?: number;
  style?: StyleProp<ViewStyle>;
  /**
   * @optional
   */
  theme: Theme;
};

/**
 * Avatars can be used to represent people in a graphical way.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img class="medium" src="screenshots/avatar-image.png" />
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Avatar } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Avatar.Image size={24} alt="avatar" source={require('../assets/avatar.png')} />
 * );
 * ```
 */
class AvatarImage extends React.Component<Props> {
  static displayName = 'Avatar.Image';

  static defaultProps = {
    size: defaultSize,
    alt: '',
  };

  render() {
    const {
      alt = '',
      size = defaultSize,
      source,
      style,
      theme,
      ...rest
    } = this.props;
    const { colors } = theme;

    const { backgroundColor = colors.primary } =
      StyleSheet.flatten(style) || {};

    return source ? (
      <View
        style={[
          {
            width: size,
            height: size,
            borderRadius: size / 2,
            backgroundColor,
          },
          style,
        ]}
        {...rest}
      >
        <Image
          source={source}
          style={{ width: size, height: size, borderRadius: size / 2 }}
        />
      </View>
    ) : (
      <AvatarText size={size} label={alt} />
    );
  }
}

export default withTheme(AvatarImage);
