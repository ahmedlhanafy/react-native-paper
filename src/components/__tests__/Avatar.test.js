/* @flow */

import * as React from 'react';
import renderer from 'react-test-renderer';
import * as Avatar from '../Avatar/Avatar';

it('renders avatar with text', () => {
  const tree = renderer.create(<Avatar.Text label="JD" />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders avatar with text and custom size', () => {
  const tree = renderer.create(<Avatar.Text size={96} label="JD" />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders avatar with text and custom background color', () => {
  const tree = renderer
    .create(<Avatar.Text backgroundColor="#FF0000" label="JD" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders avatar with text and custom colors', () => {
  const tree = renderer
    .create(
      <Avatar.Text backgroundColor="#FF0000" color="#FFFFFF" label="JD" />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders avatar with icon', () => {
  const tree = renderer.create(<Avatar.Icon icon="info" />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders avatar with image', () => {
  const tree = renderer
    .create(<Avatar.Image source={{ src: 'avatar.png' }} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
