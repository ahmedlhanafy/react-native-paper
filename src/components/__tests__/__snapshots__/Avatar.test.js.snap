// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders avatar with icon 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      },
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 64,
        },
      ]
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, .54)",
            "fontSize": 48,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders avatar with icon and custom background color 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FF0000",
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      },
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 64,
        },
      ]
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, .54)",
            "fontSize": 48,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders avatar with image 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      },
      undefined,
    ]
  }
>
  <Image
    source={
      Object {
        "src": "avatar.png",
      }
    }
    style={
      Object {
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      }
    }
  />
</View>
`;

exports[`renders avatar with text 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      },
      Object {},
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "Helvetica Neue",
          "textAlign": "left",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "rgba(0, 0, 0, .54)",
            "fontSize": 32,
            "height": 64,
            "lineHeight": 64,
          },
        ],
      ]
    }
  >
    JD
  </Text>
</View>
`;

exports[`renders avatar with text and custom background color 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FF0000",
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      },
      Object {},
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "Helvetica Neue",
          "textAlign": "left",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "rgba(0, 0, 0, .54)",
            "fontSize": 32,
            "height": 64,
            "lineHeight": 64,
          },
        ],
      ]
    }
  >
    JD
  </Text>
</View>
`;

exports[`renders avatar with text and custom colors 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 32,
        "height": 64,
        "width": 64,
      },
      Object {},
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "Helvetica Neue",
          "textAlign": "left",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "#ffffff",
            "fontSize": 32,
            "height": 64,
            "lineHeight": 64,
          },
        ],
      ]
    }
  >
    JD
  </Text>
</View>
`;

exports[`renders avatar with text and custom size 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 48,
        "height": 96,
        "width": 96,
      },
      Object {},
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "Helvetica Neue",
          "textAlign": "left",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "rgba(0, 0, 0, .54)",
            "fontSize": 48,
            "height": 96,
            "lineHeight": 96,
          },
        ],
      ]
    }
  >
    JD
  </Text>
</View>
`;
